el-patch.el

```diff
@@ -536,38 +547,78 @@ PATCH-DEFINITION is an unquoted list starting with `defun',
   (let ((definition (el-patch--resolve-definition patch-definition t)))
     ;; Then we parse out the definition type and symbol name.
     (cl-destructuring-bind (type name . body) definition
-      (let ((register-patch
-             `(let ((table (or (bound-and-true-p el-patch--patches)
-                               (make-hash-table :test #'eq))))
-                (setq el-patch--patches table)
-                (setq table
-                      (puthash ',name
-                               (gethash
-                                ',name table
-                                (make-hash-table :test #'eq))
-                               table))
-                (setq table
-                      (puthash ',type
-                               (gethash
-                                ',type table
-                                (make-hash-table :test #'eq))
-                               table))
-                (puthash el-patch-variant ',patch-definition table))))
+      (let* ((advise (and el-patch-use-advice
+                          ;; Only advice functions
+                          (let* ((props (alist-get type
+                                                   el-patch-deftype-alist))
+                                 (classifier (plist-get props :classify)))
+                            (and classifier
+                                 (equal
+                                  (caar (funcall classifier definition))
+                                  'function)))
+                          ;; Patches must have the same name and
+                          ;; same number of arguments
+                          (let ((orig-def (el-patch--resolve-definition
+                                           (cl-subseq patch-definition 0 3)
+                                           nil)))
+                            ;; Same name and same argument count
+                            (and (equal name (nth 1 orig-def))
+                                 (equal (length (nth 2 definition))
+                                        (length (nth 2 orig-def)))))
+                          'advice))
+             (register-patch
+              `(let ((table (or (bound-and-true-p el-patch--patches)
+                                (make-hash-table :test #'eq))))
+                 (setq el-patch--patches table)
+                 (setq table
+                       (puthash ',name
+                                (gethash
+                                 ',name table
+                                 (make-hash-table :test #'eq))
+                                table))
+                 (setq table
+                       (puthash ',type
+                                (gethash
+                                 ',type table
+                                 (make-hash-table :test #'equal))
```

Comment:
I had to change the test function for the variant hashtable. Nevertheless, we should check that `el-patch-variant` is not a cons or a string.